src/OpenTK.Mathematics/Matrix/Matrix3.cs

```diff
@@ -804,116 +804,52 @@ public static void Mult(in Matrix3 left, in Matrix3 right, out Matrix3 result)
         /// <summary>
         /// Calculate the inverse of the given matrix.
         /// </summary>
-        /// <param name="mat">The matrix to invert.</param>
+        /// <param name="matrix">The matrix to invert.</param>
         /// <param name="result">The inverse of the given matrix if it has one, or the input if it is singular.</param>
         /// <exception cref="InvalidOperationException">Thrown if the Matrix3 is singular.</exception>
-        public static void Invert(in Matrix3 mat, out Matrix3 result)
+        public static void Invert(in Matrix3 matrix, out Matrix3 result)
         {
-            int[] colIdx = { 0, 0, 0 };
-            int[] rowIdx = { 0, 0, 0 };
-            int[] pivotIdx = { -1, -1, -1 };
-
-            float[,] inverse =
-            {
-                { mat.Row0.X, mat.Row0.Y, mat.Row0.Z },
-                { mat.Row1.X, mat.Row1.Y, mat.Row1.Z },
-                { mat.Row2.X, mat.Row2.Y, mat.Row2.Z }
-            };
-
-            var icol = 0;
-            var irow = 0;
-            for (var i = 0; i < 3; i++)
-            {
-                var maxPivot = 0.0f;
-                for (var j = 0; j < 3; j++)
-                {
-                    if (pivotIdx[j] != 0)
-                    {
-                        for (var k = 0; k < 3; ++k)
-                        {
-                            if (pivotIdx[k] == -1)
-                            {
-                                var absVal = Math.Abs(inverse[j, k]);
-                                if (absVal > maxPivot)
-                                {
-                                    maxPivot = absVal;
-                                    irow = j;
-                                    icol = k;
-                                }
-                            }
-                            else if (pivotIdx[k] > 0)
-                            {
-                                result = mat;
-                                return;
-                            }
-                        }
-                    }
-                }
-
-                ++pivotIdx[icol];
-
-                if (irow != icol)
-                {
-                    for (var k = 0; k < 3; ++k)
-                    {
-                        var f = inverse[irow, k];
-                        inverse[irow, k] = inverse[icol, k];
-                        inverse[icol, k] = f;
-                    }
-                }
-
-                rowIdx[i] = irow;
-                colIdx[i] = icol;
+            // Original implementation can be found here:
+            // https://github.com/niswegmann/small-matrix-inverse/blob/6eac02b84ad06870692abaf828638a391548502c/invert3x3_c.h
+            Matrix3 mat = matrix;
```

Comment:
Did you try loading all matrix elements into individual locals directly instead of storing the matrix like this? I imagine it'd be much faster especially if you consider that both x64 and ARM CPUs should have enough registers to fit everything without stack spill.

Reply:
I just ran some benchmarks and it turns out making these into `Vector3`s loads them into registers, although with suboptimal code gen. With straight `float`s they get put into registers with good code gen, which is reflected in the benchmark results
```
|                Method |     Mean |    Error |   StdDev | Ratio | Code Size |
|---------------------- |---------:|---------:|---------:|------:|----------:|
|                Invert | 14.30 ns | 0.153 ns | 0.143 ns |  1.00 |     663 B |
|    InvertLocalVectors | 13.81 ns | 0.108 ns | 0.096 ns |  0.97 |     619 B |
|          InvertLocals | 10.98 ns | 0.116 ns | 0.103 ns |  0.77 |     613 B |
```